### ISSUE TITLE
Marker calulations are heavy

### ISSUE THREAD

**Issue by TheLastGimbus** [fleaflet/flutter_map#958]
Hi there!
For quite a time, I'm reaserching performance of `flutter_map` when throwing a lot of markers at it 

Recently, I created a library for drawing markers on `Canvas` instead of widgets, and it's performing pretty well. You can check it out at `dev` branch (tho it uses `flutter_map` fork for now :grimacing: ) https://github.com/KanarekApp/flutter_map_fast_markers/tree/dev

But it turns out, that, even if you don't draw makers *at all*, the calulations that translate `LatLng` to `X Y` position on the screen are taking me to <30 fps @5k markers (when not drawing them at all!)

I copy-pasted the calulations from original markers, but I can modify them - however, I have no idea how.

## Does anyone have an idea how they could be optimized?

**Comment by ibrierley** [comment]
First pondering would be to see if anyone has done similar with Leaflet.js maybe ?

**Comment by TheLastGimbus** [comment]
No dear bot, they are still heavy indeed

**Comment by aytunch** [comment]
Would using clustering help since it would decrease the number of markers being rendered on the screen?
Also if a marker is not in bounds, they are not rendered right?

**Comment by TheLastGimbus** [comment]
> using clustering

Yeah I need to check it out. Tho the existing library for them looks complicated :point_right: :point_left:

> not in bounds, they are not rendered right?

Yup. Tho:

>  the calulations that translate LatLng to X Y position on the screen are taking me to <30 fps @5k markers (when not drawing them at all!) ~Me

The caluculations that *are required* to know if marker is visible are heavy themselves

**Comment by ibrierley** [comment]
Thinking out loud, a few random bits....I have no idea if this makes sense yet to me, let alone others... :D. Suppose you were to take the highest tile zoom level eg 17, and calculate at first time only which level 17 tile it would be contained within (I guess be careful of markers that overlap tiles though..). 

If we know a marker belongs to tile x/y/z = 1024/30/17   we should be able to calculate "fairly" fast if that tile is contained within a certain range at zoom level 12 (eg 356-358/12-14/16) or whatever (as each tile zoom just raises a power of 2)? 

I'm not quite sure if it makes sense, and if it would be faster than the projection calcs, but I think you'd only need to figure it out on a tile change (ie panning within same set of tiles wouldn't need "any" recalculations at all). 

Basically like a grid eliminator, similar to geohashing (eg https://www.ibm.com/docs/en/db2/11.5?topic=concepts-geohashes-geohash-covers ). I guess I'm trying to figure if a tile is sort of like a geohash (or are the tiles redherrings and could simply use geohashing ?) Something a bit like this... https://stackoverflow.com/questions/36705355/finding-geohashes-of-certain-length-within-radius-from-a-point.

Again, sorry it's a bit random, just prodding a few thoughts.

**Comment by ibrierley** [comment]
In fact, I think the tile stuff is indeed a redherring...and maybe geohashing the markers at first, then on an geohash area change, recheck them which could eliminate a lot. It's not trivial, but could be interesting, just depends if it's worth the effort to investigate. There's a dart geohash lib https://pub.dev/packages/dart_geohash...

So for example if we look at the pic on https://gis.stackexchange.com/questions/231719/calculating-optimal-geohash-precision-from-bounding-box

lets say we have a point which is geohash "gce46hy7" and we are looking roughly at a map of the UK, you may test points that start with the strings  gf or gc or gb or u0, u1, u4.... so "gce46hy7" starts with "gc" so is inside and passes the test. If we have a point "eg5qr5y" it starts with "eg" which isn't part of that set, so would be excluded.

Not sure if I'm barking up the wrong tree, or if its slower, but again, I think you would only need to do a test when the map moves across to a new set of geohash strings.

Clear as mud I hope :D

**Comment by TheLastGimbus** [comment]
Still an issue :confused:

**Comment by Y0ngg4n** [comment]
can i use the https://github.com/KanarekApp/flutter_map_fast_markers/tree/dev allready or is it under development currently?

**Comment by TheLastGimbus** [comment]
@Y0ngg4n We use it at Canary app and so far been *very* satisfied - if you avoid heavy drawings like shadows and stick to simple shapes like circles etc you can have ~4k markers on most phones without bigger problems :+1: 

The API of how you do the drawing *may* change, but will still rely on canvas (99% same)

**Comment by Y0ngg4n** [comment]
@TheLastGimbus the fork seems to be some commits behind. is it possible to update it to the latest? 
Can i use it with the cluster plugin?

**Comment by Y0ngg4n** [comment]
I have tried with clustering. But it seems to make no difference if using clustering or not 🤔.

**Comment by TheLastGimbus** [comment]
I will kill you. No, do not close this.

**Comment by ibrierley** [comment]
So where is this currently, doesn't the fast_markers get around the problem by caching and not having to recalculate every frame ? Just trying to remember what's outstanding on the issue.

**Comment by TheLastGimbus** [comment]
I think everything what i said above stands. All cache optimizations that I made were already merged here in #826 - everything is sitll heavy even with them

**Comment by ibrierley** [comment]
It's not so much the if the above stands, just trying to understand whereabouts the problem is.

Looking at the fastmarker code, you don't calculate the projection unless the zoom is switched (as using the pxCache) I think ? So which bit is the heavy calculation you are looking at ?

**Comment by JaffaKetchup** [comment]
Been following this for a while, but haven't said anything due to lack of experience. But here's my thoughts:

I was looking at your libraries code, and found this: `map.project(marker.point);`, which I'm assuming is the calculations you mentioned in your original body?

Having recent experience with this, the projection function is actually just a call to another function which is split into 3 parts, which you can look up here if interested:
https://github.dev/fleaflet/flutter_map/blob/95a8d26b353fc3cc62f29f210c2a91be8a5a2d5f/lib/src/geo/crs/crs.dart#L26.

Without digging extremely deep into `flutter_map` and some of it's dependencies, it's not clear which part is slow. Finding which part is slow - either the actual projection, scaling, or transformation - may mean we can actually focus on the calculations instead of finding workarounds, like caching, or one I'm about to suggest.

The issue is, those functions depend on the current CRS and projection, and also reference to the external library `proj4dart`. This makes them way to complicated for me, and perhaps altogether out of scope, but someone can have a further look if they want. The bigger issue is that the maths needed will be complicated, and there's no way around that. But those calculations are currently performed blocking the main thread.

So here's my workaround suggestion, which can be used in conjunction with caching or not: isolates.
Yes, I know they are unpopular and difficult, but [Dart 2.15 (video changelog)](https://www.youtube.com/watch?v=UKH9ulkribk&ab_channel=FilipHr%C3%A1%C4%8Dek) has just been released, and with it, [major changes to isolates](https://medium.com/dartlang/dart-2-15-7e7a598e508a#:~:text=publishers%2C%20and%20more.-,Fast%20concurrency%20with%20worker%20isolates,-Just%20about%20all) to make them faster and easier to code (open the latter link in Chrome for automatic highlighting of the correct passage).

Perhaps just wrapping the calls in isolates will improve the speed enough. Alternatively, this could just make it worse.

As I said, I don't know how to do much more than this, so this is as much as I can suggest. Hope it helps, and thanks for your great work already!

**Comment by ibrierley** [comment]
Nothing wrong with your suggestions! Isolates could help with a large volume of projections when zoom switching if that's the issue.

The thing is, there are different problems here, and it's not clear (to me) atm. I ran the example code, which draws 5,000 markers on the screen.

If I comment out the 2 canvas.drawLine calls, I get 60 fps. Unless I switch over a zoom boundary when it makes the conversions just for a moment (and if thats the issue, I guess one could have a cache per zoom level).

If I leave them in, I get 60 fps when markers off screen (displayed but out of bounds), 8 fps when all 5000 on display (which I assume will never happen in real life). 

There are possibly ways to optimise the canvas calls, but I assume that's not the issue here, as I assume this is just a test example. 

So, like I said, I'm just trying to understand the nature of the problem. map.project is only called on a level switch.

So is the problem...
a) it's slow all the time, even when panning and markers not in view (I don't see this, but it looks like it's just doing basic boundary checks, not projections. There could be optimisations if for example there are 1 million objects but only a small number in view)
b) it's slow when zooming and switching to a new zoom level (this seems to be down to project calls, maybe isolates could help, or some faster algorithm etc, or cache for each zoom level).
c) it's slow when there's a lot of markers in view (this seems to be down to canvas calls, and there could be optimisations).
d) some other issue or combo

Depending upon the exact nature of the problem, would vary the solution.

**Comment by JaffaKetchup** [comment]
I think b is probably the thing we should focus on for now? But not sure.

@TheLastGimbus any ideas?

**Comment by TheLastGimbus** [comment]
I'm not currently at work with this, so I can't analyze this broadly right now :( But from what I see, having another Isolate may be a necessary evil at the moment :(

**Comment by JaffaKetchup** [comment]
Marking as high-priority for now...

**Comment by mohammedX6** [comment]
I am wondering if I can combine `fast_marker_plugin` with `flutter_map_marker_cluster` the result would be good?

**Comment by ibrierley** [comment]
Like I mentioned earlier. I think it really needs people to be clear about the specific problem they are having each time. How many markers, over what area etc. With 5000 it was fine for me (some slowdown if all 5000 were on screen, but are people really displaying that many?)

**Comment by mohammedX6** [comment]
I have a really complex app with around 8 layers stacked on each other (Polygon,Polyline,Points), I experience lags when I am using clustering for markers finished

**Comment by ibrierley** [comment]
How many markers are there and how are you clustering them  ?

**Comment by aytunch** [comment]
@mohammedX6 Can you give out some more details on the number of total&onscreen markers please?

**Comment by mohammedX6** [comment]
@ibrierley  @aytunch  I have 3 `flutter_map_marker_cluster`  layers, all contain around 4000 markers

**Comment by ibrierley** [comment]
@TheLastGimbus will probably disagree, but I feel like this should be closed without more info now. It's getting a bit conflated with the original problem, and open a new issue with the specific area that's causing a problem (or add more clarity to this issue). 

From what I can tell (and may well be wrong), markers are cached using _pxCache in the code, so it needs clarity what exactly the problem is since that caching was introduced. I don't think things like isolates should be introduced into flutter_map code (yet), but could of course be part of any plugin, which I feel should be the way to go.

I am very open to options though, just wary about this ever getting anywhere!

**Comment by felix-ht** [comment]
In the end what is needed is to break the O(n) barrier. If you always have to check everything even though its not visible, performance will always degrade linearly with marker (feature) count.

A good solution for this would probably be to automatically split the markers into tiles. Then checking what to draw is as simple as calculating what tiles should be currently be drawn in the current zoom level, and extracting the markers from that tile. This has the effect that only features on the currently visible tile would ever be considered for drawing - no need to go through all of them every time.

If combined with clustering one can pretty much keep the actual number of markers that have to be considered for drawing constant, regardless if there are 200 or 200000 on the map. On lower zoom levels and in locations with high density the markers would get clustered and on higher ones the number of visible ones decreases. This is possible because the tiles don't not have to hold the markers directly, but their representation as a clustered point (with metadata). This would also require separate to concerns of where to draw things (geometry+metadata) from how draw them (calling a draw function for each visible geometry - _not_ necessarily the original markers)

The ability to have multiple differing representations of the same object across the zoom range is the biggest advantage of tiling, over "just" going with some other space partitioning approaches.

Other spatial indexes (hashing, quadtrees, S2, hilberscuves and so on) would work as well, but you still have to deal with all the markers (features) as you zoom out. So the performance would still degrade on these zoom levels.

**Comment by Y0ngg4n** [comment]
anything new?

**Comment by ibrierley** [comment]
What problem are you having atm, with the caching, it's pretty decent. I tried the other day with 5,000 markers. How many do you want to display? :)

**Comment by Y0ngg4n** [comment]
about 60-100k

**Comment by ibrierley** [comment]
Can you explain your use case out of interest, and why wouldn't you use clustering for example for that many ?

**Comment by Y0ngg4n** [comment]
I use clustering, but clustering does not reduce the lag.

**Comment by ibrierley** [comment]
So how many markers would typically be displayed on your screen with clustering ?

It feels like there's 2 things here, so trying to be careful not to get them mixed up....calculations which a plugin may use (eg cluster plugin, and really that's down to the plugin to resolve) and the flutter_map markers which can use caching.

**Comment by ibrierley** [comment]
Just to clarify a bit that. I think the key is, the calculations aren't likely to change much. The key imo is not to do the calculations in the first place (by caching, grouping, breaking down into areas etc, well you probably need to do them once at least, but not every frame).

**Comment by Y0ngg4n** [comment]
With clustering around 200-500. But clustering did not helped. It lagged the same as before.

**Comment by ibrierley** [comment]
Do you have some example test data for the 100k or so markers ?

**Comment by ibrierley** [comment]
So, out of interest, I hacked up an example I did with a tile slicing proof of concept plugin I did, and tiled clusters. I just wanted to get an idea of what could be possible if energy was put into it. It's not production ready (there are some flaws in this method), and probably never will be if it's just left to me :D, but I would be happy to help out with the concept if anyone gets interested. 

This uses 100,000 markers saved into a geojson map (it fell over at 1 million, but there may be optimisations to be done). Note, it's a bit slow for the first few seconds whilst calculating and downloading new tiles, and icons aren't properly centred etc.

https://youtube.com/shorts/mI9g-Kvkv64

**Comment by Y0ngg4n** [comment]
no i fetch them from the database. 
I have done a workaround to only fetch the points that are in a certain range to fix it. but i am not quite happy with it.

**Comment by TheLastGimbus** [comment]
Hi there dear bot, thanks for checking up :heart: I'm good, thanks for asking. My vacation is going great so far, though it keeps raining :confounded: but the weather ain't a problem if you have so many good friends (like you :eyes:). How about you? Are the kids doing well? 

Take care! Bye!

**Comment by JaffaKetchup** [comment]
@TheLastGimbus Chatting to the bot again I see :D.

We haven't forgotten about this, we're just not sure what to do! Also I think @ibrierley has those projects he mentioned above, but not sure how far along those are?

**Comment by JaffaKetchup** [comment]
Bot says:

```
My vacation is going stale because it has also been raining here for 30 days. If it doesn't stop within 5 days, I might have to stop the vacation, and come and visit you (:)). Children are doing very well thank you.

Enjoy yourself!
```

_Anyone reading in on this, don't worry about it. None of us have gone insane._

**Comment by ibrierley** [comment]
If anyone has a specific problem they want to create a minimal example of, I suspect it will get more traction. Currently it's hard to know who is using clusters, who is trying to get 100,000 widgets (of what?) displayed on zoom level 0, who is trying to get 200 displayed spread across the map. Are the markers static or dynamic (i.e how much caching can be used etc). So realistically as I've said before, I don't think this issue is really going anywhere without more focus (as it currently is). If they are using FM Markers and it's using caching, how are the calculations heavy etc.

**Comment by MooNag** [comment]
I did some performance benchmarks on this branch #1333 and my conclusion is that out of the rendering time, bounds checking is small. On my Note 20 on the example app projecting and bounds checking the 5,000 points takes only 5ms total (200 fps), often less, out of a 17ms frame time (55 fps).
Basically by far the heaviest part of displaying 5000 markers is rendering them to the screen, not checking if they are there. Your issue is most likely the **markers you are trying to render are simply too heavy. That isn't something that can be optimized other than not rendering all of them.**

If you need to render more than 5k or so markers then you need to cluster them. I cannot say anything about the clustering plugin performance, but my guess is if it is properly optimized it could handle in the range of 50k points on mobile.

This issue should be closed and re-opened on the clustering plugin if you are having performance issues with clustering.

**Comment by JaffaKetchup** [comment]
Ok, I'm going to close this for now. There might be some improvements coming from #1333 in the near future.

**Comment by rorystephenson** [comment]
@Y0ngg4n and @TheLastGimbus I'm not sure where you ended up with these performance issues but I just wanted to let you know that I've been working on a [new clustering layer](https://pub.dev/packages/flutter_map_supercluster) which is, at least in my testing, very fast. I'm currently working with around 10,000 points and they are reasonably clustered so I'd be curious to know how this performs for you with more points.

If there are still performance issues then the next step would be using a painter to render the markers rather than using widgets. There would be a couple of tradeoffs like not being able to maintain state for markers and needing to add some hit-testing code for marker taps but I think there would be a pretty big difference in performance.

I should also mention that the indexes which are used to make the clustering package fast could be used for a plain marker layer asa well if you are not interested in clustering. I'm happy to share an example if you like.

**Comment by aytunch** [comment]
@rorystephenson thanks for this plugin.

> There would be a couple of tradeoffs like not being able to maintain state for markers and needing to add some hit-testing code for marker taps

For my usecase, dynamic widget markers are a must. So I hope the performance is good even without custom paint.

**Comment by rorystephenson** [comment]
> @rorystephenson thanks for this plugin.
> 
> > There would be a couple of tradeoffs like not being able to maintain state for markers and needing to add some hit-testing code for marker taps
> 
> For my usecase, dynamic widget markers are a must. So I hope the performance is good even without custom paint.

Let me know how it goes! If it's still not performing well I'd be happy to try to figure out why.

**Comment by TheLastGimbus** [comment]
> using a painter to render the markers rather than using widgets

Yesss, that's what https://github.com/KanarekApp/flutter_map_fast_markers/ are for. We didn't migrate the library for new version *yet*, but once we do we'll let know how it performs!

**Comment by JaffaKetchup** [comment]
Hey @TheLastGimbus 👋. Would you like me to add your plugin to the docs?

**Comment by TheLastGimbus** [comment]
@JaffaKetchup it would be great!! However, this plugin is currently not migrated to 3.0 at all :/ i will get in touch when I figure it out, okay?

@Pepslee yes, i want it to be a plugin in future. Yes, it would need some custom stuff for clusterization :/

**Comment by JaffaKetchup** [comment]
Ok @TheLastGimbus. For now I will set it as Beta until everything is sorted :)

**Comment by rorystephenson** [comment]
> @Pepslee yes, i want it to be a plugin in future. Yes, it would need some custom stuff for clusterization :/

The clustering algorithms/logic behind my [`flutter_map_supercluster`](https://pub.dev/packages/flutter_map_supercluster) plugin is contained in my [`supercluster`](https://pub.dev/packages/supercluster) plugin which may be useful for you.
